File: 프로그래머스/lv2/131127. 할인 행사/할인 행사.java

/*
    1~10일 2~11일 ... -> 슬라이딩 윈도우
    시간복잡도 = 10 * 100,1000 < 1d억
    
    걸린시간 : 18분
*/


import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        HashMap<String,Integer> map = new HashMap<>();
        int answer=0;
        
        for(int i=0;i+9<discount.length;i++){
            if(i==0){
                for(int j=0;j<10;j++){
                map.put(discount[j],map.getOrDefault(discount[j],0)+1);
                }
            }else{
                map.put(discount[i-1],map.getOrDefault(discount[i-1],0)-1);
                map.put(discount[i+9],map.getOrDefault(discount[i+9],0)+1);
            }
            
            boolean valid=true;
            
            for(int j=0;j<number.length;j++){
                if(map.get(want[j])==null||map.get(want[j]) != number[j]){
                    valid=false;
                    break;
                }
            }
            
            if(valid){
                answer++;
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
def solution(want, number, discount):
    buyDict = {};
    start = 0
    answer = 0
    
    while start + 9 < len(discount):
        if start == 0:    
            curr = 0
            
            while curr < 10:
                buyDict[discount[curr]] = buyDict.get(discount[curr], 0) + 1
                curr = curr + 1
        else:
            buyDict[discount[start - 1]] = buyDict.get(discount[start - 1], 0) - 1
            buyDict[discount[start + 9]] = buyDict.get(discount[start + 9], 0) + 1
        
        success = True
        
        for i in range(0,len(want)):
            if buyDict.get(want[i],0) < number[i]:
                success = False
                break
        
        if success == True:
            answer = answer + 1
        
        start = start + 1
    
    return answer